Add support for full ViT optimization

iAnt0n opened the pull request:
### Description
<!-- Describe your changes. -->
Add support for ViT optimization in optimizer.py
As ViT architecture follows BERT rather closely, we can easily reuse BERT fusions for ViT. The only difference is that ViT does not have attention mask, which means there is no Add node in qk paths.
Make the necessary changes in onnx_exporter.py to be able to cover optimizations with test.


### Motivation and Context
<!-- - Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here. -->
Vision transformer optimizations are not supported yet, and ViT is pretty straightforward to add. This can be a good first step to have more officially supported models.

iAnt0n commented:
@microsoft-github-policy-service agree

tianleiwu commented on onnxruntime/python/tools/transformers/fusion_attention.py:
```diff
@@ -458,6 +458,16 @@ def fuse(self, normalize_node, input_name_to_nodes, output_name_to_node):
 
+        children = input_name_to_nodes[root_input]
+        children_types = [child.op_type for child in children]
+        if children_types.count("MatMul") != 3:
```
Add some comments for this block of code like example graph pattern.

iAnt0n replied:
Done, thanks
I also feel like we can remove [these lines of code](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/python/tools/transformers/fusion_attention.py#L427-L435) completely, since this case with multiple Add consumers is handled by SkipLayerNorm fusion itself since [this commit](https://github.com/microsoft/onnxruntime/commit/7ed8bd4f95977d7601250564e3f40039d176a2c7#diff-2c80d28126f62441ef0d93ae1c9021a1c771d24bb9d4737feb606f45dd45c1ff)

tianleiwu replied:
From my understanding, this code is to find root_input for SkipLayerNormalization.

It may be better to move to a place that computes root_input.

iAnt0n replied:
It is not about normalize_node, but about its parent. When its parent is usual SkipLayerNorm, it has the same output for Attention and this normalize node, but when it is a SkipLayerNorm with multiple Add consumers, it has two outputs - one goes to Attention, and the other goes to normalize_node. In this case we need to pick the other SkipLayerNorm output.

Actually, having looked at this once again, I rewrote this part of code so that it is more explicit and concise, thanks

tianleiwu commented:
Related PR: https://github.com/microsoft/onnxruntime/pull/15045

tianleiwu commented on onnxruntime/python/tools/transformers/requirements.txt:
```diff
@@ -8,6 +8,7 @@ packaging
 transformers >= 4.18.0
 scipy
 sentencepiece
+pillow
```
Is the pillow package needed for test? I did not see the import in script.

iAnt0n commented on onnxruntime/python/tools/transformers/fusion_attention.py:
```diff
@@ -458,6 +458,24 @@ def fuse(self, normalize_node, input_name_to_nodes, output_name_to_node):
+        """
+        parent_node = output_name_to_node[root_input]
+        if parent_node.op_type == "SkipLayerNormalization" and len(parent_node.output) == 4:
+                root_input = parent_node.output[0]
```
Done, thanks

tianleiwu commented:
/azp run Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,ONNX Runtime Web CI Pipeline

tianleiwu commented:
/azp run Windows ARM64 QNN CI Pipeline

tianleiwu commented:
@iAnt0n, it seems that some pipeline failed:
`` ImportError: cannot import name 'AutoImageProcessor' from 'transformers'``

It is because the pipeline use transformers 4.24, and I think AutoImageProcessor need 4.26 or later.
https://github.com/microsoft/onnxruntime/blob/e4aae94f20b995fe4b80643499d46af211e38fc2/tools/ci_build/requirements.txt#L6

You can try install the requirements.txt, and run `pytest` under `onnxruntime\test\python\transformers` directory to see whether the test can pass.

iAnt0n commented:
> Another possible walkaround is to move `import AutoImageProcessor` to right before the line that uses the class. Since the test for vit is marked as slow, so the test will not run in pipeline.

I think it won't be nice to have slow tests require other versions of libraries, even though they are not running in CI.
ImageProcessor classes came to replace FeatureExtractor in [Transformers 4.25.1](https://github.com/huggingface/transformers/releases/tag/v4.25.1#:~:text=in%20%2317799-,Image%20processors,-Image%20processors%20replace), and they are backwards compatible, which means we can use AutoFeatureExtractor here instead of AutoImageProcessor.

Changed AutoImageProcessor to AutoFeatureExtractor, the test passes with Transformers 4.24. Updated PR.

tianleiwu commented:
/azp run Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,ONNX Runtime Web CI Pipeline

tianleiwu commented:
/azp run Windows CPU CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline,onnxruntime-binary-size-checks-ci-pipeline,onnxruntime-python-checks-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,Windows ARM64 QNN CI Pipeline

tianleiwu commented:
@iAnt0n, thanks for the contribution
